abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+  is paused from it).
+- `HTTP_RESPONSE_BODY` can be read and modified in
+  [`proxy_on_response_body`] (or for as long as response processing
+  is paused from it).
+- `DOWNSTREAM_DATA` can be read and modified in
+  [`proxy_on_downstream_data`] (or for as long as upstream processing
+  is paused from it).
+- `UPSTREAM_DATA` can be read and modified in
+  [`proxy_on_downstream_data`] (or for as long as downstream processing
+  is paused from it).
+- `HTTP_CALL_RESPONSE_BODY` can be read in
+  [`proxy_on_http_call_response`].
+- `GRPC_CALL_MESSAGE` can be read in [`proxy_on_grpc_receive`].
+- `VM_CONFIGURATION` can be read in [`proxy_on_vm_start`].
+- `PLUGIN_CONFIGURATION` can be read in [`proxy_on_configure`].
+- `FOREIGN_FUNCTION_ARGUMENTS` can be read in
+  [`proxy_on_foreign_function`].
+
+
+### Functions exposed by the host
+
+#### `proxy_set_buffer_bytes`
+
+* params:
+  - `i32 (`[`proxy_buffer_type_t`]`) buffer_id`
+  - `i32 (uint32_t) offset`
+  - `i32 (uint32_t) size`
+  - `i32 (const uint8_t *) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sets content of the buffer `buffer_id` to the provided value
+(`value_data`, `value_size`) replacing `size` bytes starting
+at `offset` in the existing buffer.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `buffer_id`.
+- `NOT_FOUND` when requested `buffer_id` isn't available.
+- `INVALID_MEMORY_ACCESS` when `value_data` and/or `value_size`
+  point to invalid memory address.
+
+
+#### `proxy_get_buffer_bytes`
+
+* params:
+  - `i32 (`[`proxy_buffer_type_t`]`) buffer_id`
+  - `i32 (uint32_t) offset`
+  - `i32 (uint32_t) max_size`
+  - `i32 (uint8_t **) return_value_data`
+  - `i32 (size_t *) return_value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves up to` max_size` bytes starting at `offset` from the buffer
+`buffer_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `buffer_id`, or in case of buffer overflow
+   due to invalid `offset` and/or `max_size` values.
+- `NOT_FOUND` when requested `buffer_id` isn't available.
+- `INVALID_MEMORY_ACCESS` when `returned_value_data` and/or
+  `returned_value_size` point to invalid memory address.
+
+
+#### `proxy_get_buffer_status`
+
+* params:
+  - `i32 (`[`proxy_buffer_type_t`]`) buffer_id`
+  - `i32 (size_t *) return_buffer_size`
+  - `i32 (uint32_t *) return_ignored`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves size (`return_buffer_size`) of the buffer `buffer_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `buffer_id`.
+- `NOT_FOUND` when requested `buffer_id` isn't available.
+- `INVALID_MEMORY_ACCESS` when `return_buffer_size` and/or
+  `return_ignored` point to invalid memory address.
+
+
+## HTTP fields
+
+Access to HTTP fields using functions in this section is restricted to
+specific callbacks:
```

Comment:
add reference to `proxy_header_map_type_t` to indicate that the following items correspond to one of the enum values?